### PR TITLE
Add Server SQLite in background

### PR DESCRIPTION
**Pull request by Shosoul**

### CODE REVIEW

**Comment by Shosoul** [comment]
Note: this pr will likely get closed and re-made for #786

**Comment by Shosoul** [comment]
> What is with the PADatabase creating all the tables and also having a way to make them with sql?
For ease of testing, It can be removed in the final version.
> Also we should import all the current players from players.txt when the tables are first created
I'll see if I can have it import into the database when the server first loads up with a loop or something.

**Comment by Shosoul** [comment]
Okay, the server now loads in users with their ratings from the leaderboard when the server starts up.

**Comment by Shosoul** [comment]
This puts SQLite in the server. The code currently works to record matches and changes and the server does not actually depend on the SQL database yet.

File: server/PADatabase.lua [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+  gameID INTEGER NOT NULL,
+  level INTEGER,
+  placement INTEGER NOT NULL,
+  FOREIGN KEY(playerID) REFERENCES Player(publicPlayerID),
+  FOREIGN KEY(gameID) REFERENCES Game(gameID)
+);
+]]
+--[[
+
+CREATE TRIGGER IF NOT EXISTS trig_ratingUpdate BEFORE UPDATE ON Player
```
[the remark this answers is not in the document]
Comment (Shosoul) [reply server/PADatabase.lua]: Done

File: server/PADatabase.lua [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+  end
+end]]
+
+local selectPlayerRecordValuesStatement = assert(db:prepare("SELECT * FROM Player where privatePlayerID = ?"))
+local function getPlayerValues(privatePlayerID)
+  selectPlayerRecordValuesStatement:bind_values(privatePlayerID)
+  selectPlayerRecordValuesStatement:step()
+  local playerValues = selectPlayerRecordValuesStatement:get_named_values()
+  if selectPlayerRecordValuesStatement:reset() ~= 0 then
+    print(db:errmsg())
```
[the remark this answers is not in the document]
Comment (Shosoul) [reply server/PADatabase.lua:54]: done

File: server/createpadatabase.sql [hunk 3]
```diff
@@ -0,0 +1,24 @@
+DROP TABLE IF EXISTS PlayerELOHistory;
```
[the remark this answers is not in the document]
Comment (Shosoul) [reply server/createpadatabase.sql]: done

File: server/PADatabase.lua (hunk 2 above)
[the remark this answers is not in the document]
Comment (Shosoul) [reply server/PADatabase.lua:54]: Oh. I must have forgotten to save the file before I committed. Whoops!

[the remark this answers is not in the document]
Comment (Shosoul) [reply server/PADatabase.lua]: Fixed.

File: server/PADatabase.lua [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+db:exec[[
+PRAGMA foreign_keys = ON;
+
+CREATE TABLE IF NOT EXISTS Player(
+  publicPlayerID INTEGER PRIMARY KEY AUTOINCREMENT,
+  privatePlayerID INTEGER NOT NULL UNIQUE,
+  username TEXT NOT NULL,
+  placementDone BOOLEAN NOT NULL CHECK (placementDone IN (0, 1)) DEFAULT 0,
+  placementRating REAL NOT NULL DEFAULT 0,
+  rankedGamesPlayed INTEGER NOT NULL DEFAULT 0,
```
Comment (Sankyr) [inline server/PADatabase.lua:20]: are these fields necessary? couldn't we get this info from joining with the PlayerGameResult & Game table and counting the ranked games & placement?

File: server/PADatabase.lua [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+)
+
+db:exec[[
+PRAGMA foreign_keys = ON;
+
+CREATE TABLE IF NOT EXISTS Player(
+  publicPlayerID INTEGER PRIMARY KEY AUTOINCREMENT,
+  privatePlayerID INTEGER NOT NULL UNIQUE,
+  username TEXT NOT NULL,
+  placementDone BOOLEAN NOT NULL CHECK (placementDone IN (0, 1)) DEFAULT 0,
```
Comment (Sankyr) [inline server/PADatabase.lua:18]: Could this be done with Endaris's suggestion of counting the number of ranked games played from the player and storing the placement rating in the ELO table?

File: server/PADatabase.lua [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+  placementRating REAL NOT NULL DEFAULT 0,
+  rankedGamesPlayed INTEGER NOT NULL DEFAULT 0,
+  rankedGamesWon INTEGER NOT NULL DEFAULT 0,
+  lastLoginTime TIME TIMESTAMP DEFAULT (strftime('%s', 'now'))
+);
+
+CREATE TABLE IF NOT EXISTS PlayerELOHistory(
```
Comment (Sankyr) [inline server/PADatabase.lua:25]: shouldn't this table also have the gameID which caused the update? (which would remove the need for updateTime)

File: server/PADatabase.lua [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+);
+
+CREATE TABLE IF NOT EXISTS Game(
+  gameID INTEGER PRIMARY KEY AUTOINCREMENT,
+  ranked BOOLEAN NOT NULL CHECK (ranked IN (0, 1)),
+  timePlayed TIME TIMESTAMP NOT NULL DEFAULT (strftime('%s', 'now'))
+);
+
+CREATE TABLE IF NOT EXISTS PlayerGameResult(
+  playerID INTEGER NOT NULL,
```
Comment (Sankyr) [inline server/PADatabase.lua:39]: change this to publicPlayerID for consistancy?

File: server/PADatabase.lua (hunk 4 above)
Comment (Shosoul) [reply server/PADatabase.lua:20]: Some of these fields are not needed, I will get rid of them

File: server/PADatabase.lua (hunk 5 above)
Comment (Shosoul) [reply server/PADatabase.lua:18]: If I end up doing placement matches, I'll do it in a separate PR

File: server/PADatabase.lua (hunk 6 above)
Comment (Shosoul) [reply server/PADatabase.lua:25]: Only thing about this is that I have no game to tie imports of people's ratings from the CSV file

File: server/PADatabase.lua (hunk 7 above)
Comment (Shosoul) [reply server/PADatabase.lua:39]: Will do.

File: server/PADatabase.lua (hunk 6 above)
Comment (Sankyr) [reply server/PADatabase.lua:21]: I would rather have them reference a null game (or some default game timestamped at the creation of the database) than to sacrifice this entire schema?

File: server/PADatabase.lua (hunk 7 above)
Comment (Shosoul) [reply server/PADatabase.lua]: Done

**Pull request merged**